system/src/Grav/Common/Page/Medium/ImageMedium.php

```diff
@@ -232,31 +218,36 @@ public function srcset($reset = true)
      * @return $this
      */
     public function derivatives($min_width, $max_width, $step = 200) {
-      $width = $min_width;
-
-      // Do not upscale images.
-      if ($max_width > $this->get('width')) {
-        $max_width = $this->get('width');
-      }
-
-      while ($width <= $max_width) {
-        $ratio = $width / $this->get('width');
-        $derivative = MediumFactory::scaledFromMedium($this, 1, $ratio);
-        if (is_array($derivative)) {
-          $this->addDerivative($derivative['file']);
+        $width = $min_width;
+
+        // Do not upscale images.
+        if ($max_width > $this->get('width')) {
+            $max_width = $this->get('width');
         }
-        $width += $step;
-      }
-      return $this;
-    }
 
-    /**
-     * Add a derivative
-     *
-     * @param  ImageMedium $image
-     */
-    public function addDerivative(ImageMedium $image) {
-      $this->derivatives[$image->url()] = $image->get('width');
+        // Clear any alternatives that have already been implied by the image's
+        // filename (eg. ones that had "@2x" or similar appended to their
+        // filenames)
+        $this->clearAlternatives();
+
+        while ($width <= $max_width) {
+            $ratio = $width / $this->get('width');
+            $width += $step;
+
+            if ($ratio == 1) {
+                continue;
+            }
+
+            $derivative = MediumFactory::fromFile($this->get('filepath'));
```

Comment:
I haven't done any benchmarking between the current implemention and this patch, but yes this could improve performance.

While `MediumFactory::scaledFromMedium` (which is used in the current implementation) performs both a write and a read in this step, this patch performs just a read. The difference is that this patch always reads the original file, while the old implementation only reads the original file once, writes it several times, and then reads the copy it wrote when adding the derivative. If the original file is very large, I can imagine that performance would be better by not reading *that* file several times, but I don't really know without measuring.